<h1>
<center>CFRM 421 Group Project


<h1>
<center>Prediction of Stock Price Changes of Dow Jones Index and Stocks</center>
</h1>

* **Group Members:
Charlie Yin,
Tianrui Zhou,
Sayako Mitchell**

* In this report, we will firstly give an introduction of our project and data set. Then, we will explain how we preprocessed our data. The existing literature and introduction of every model (as well as the algorithm) will also be covered. We will finally talk about our results and conclude our findings, and we will also point out some possible future research directions in the end.

* Every member in this group are highly envolved and we're all responsible for this project. Behind each section name, the name of the editor will be clearly listed out. *Charlie Yin*


# 1. Introduction (*Tianrui Zhou*)

Nowadays, the technology industry is a hot topic, and many tech stocks have outstanding performance among all the stocks. In this project, our primary objective is to predict the percentage change in the closing prices of  3 chosen stocks from the technology industry of the Dow Jones Index and compare the predictions with the predicted Dow Jones Index Average performance.
Each team member experiments with a different regression model— Recurrent Neural Network Model (RNN), Support Vector Machine regression(SVM), and Random Forest Regressor—to predict the percentage change in closing prices for the three top-performing tech stocks and Index Average.
The aim is to determine if specific stocks outperform the Index Average according to our predictions using three different machine learning models, and compare the result with the real value, which can reveal which model performs the best. Our methodology is structured into six comprehensive steps:

**Step 1:** Find the top 3 technology stocks that outperform the DJIA index most.

**Step 2:** For each model, hyperparameter tune only on the top outperform stock “MSFT” that we just found in step 1. Use the resulting model with the best parameter to predict the percentage change of the stock “MSFT” of Friday, May 24th.

**Step 3:** Use the best parameter, and retrain the model on the other two stocks and then do predictions like step 2.

**Step 4:** Each person uses their own model for DJIA index, do hyperparameter tuning on DJIA data, and predict the percentage change of the Index Average's closing price for the same day.

**Step 5:** For each model’s prediction result, we calculate the difference between the percentage Index Average price change and the percentage stock price change. Then, we calculate the real difference, which is the difference between the real price change of the Index Average and the real percentage change of the stock price observed on Friday, May 24th. We first check if the predicted difference and the real difference are in the same sign (positive or negative), which indicates if a stock outperforms the Index Average. Second, for each stock, we calculate the error, which is the absolute value of the difference between the real difference and the predicted difference, in order to evaluate the accuracy of each model.

**Step 6:** For choosing the best model, we consider the difference between the real price change and the predicted price change Select the best model among the three based on the performance on step 5.

# Existing Literature

A large number of papers have been published on the topic of predicting the performance of stocks. Many of them use the same methods we will employ in this project.

***Neural Network (Charlie Yin)***

My insights into building the RNN-LSTM model are from the paper named” Stock Price Prediction using the RNN Model”  (https://iopscience.iop.org/article/10.1088/1742-6596/1650/3/032103/pdf )  by Yongqiong Zhu.  In this paper, Yongqiong built an RNN model with 2 layers and got great stock prediction performance. Our project is predicting the percentage change of price, which is slightly different, but this paper enlightens me on the application of RNN and its variants in prediction tasks. Combining what I learned from this class and from the paper by Yongqiong Zhu, I decided to try the RNN-LSTM model in our project.

***Random Forest (Saya Mitchell)***

Random Forest Regression is used to predict stock trends using historical data in papers such as “The Random Forest Model for Analyzing and Forecasting the US Stock Market in the Context of Smart Finance” (https://arxiv.org/pdf/2402.17194) by Jiajian Zhen, Duan Xin, Qishuo Chen, Miao Tian, and Le Yang. This paper found that random forest models can generate up to 85 to 95 percent accurate predictions. It should be noted that the authors of this paper used data preprocessing methods that are outside of the scope of this class, including exponential smoothing. However, the core tasks of this paper and our project are similar. Additionally, this project used data from technology and technology-adjacent stocks (Apple, Samsung, and General Electric), which further aligns with our project.

***Support Vector Machines (Tianrui Zhou)***

I encountered a paper called “Stock Price Prediction Using Support Vector Machine Approach” by Naliniprava Tripathy (https://www.dpublication.com/wp-content/uploads/2019/11/24-ME.pdf). In this paper, they are trying to predict the closing price of S&P BSE TECK index, and the resulting accuracy of the SVM regression model is about 60.2%. In this case, our project has similar goal. Therefore, SVM regressor may be a potential nice model to predict the stock price.


# Methods

**Neural Network**

Recurrent Neutral Network (RNN) model is a deep learning model that is trained to process and convert a sequential data input into a specific sequential data output. Time series data is one of the most frequently used data as “sequential data” in RNN model. Recurrent neural networks produce predictive results in sequential data that other algorithms can’t.
Long Short-Time Memory Model can recall previous long-term time-series data and has automatic control for retaining relevant features or discarding irrelevant features in the cell state.

In my model, I add one layer of Long Short-Time Memory layer with ReLu as an activation function to help the RNN model in maintaining and utilizing long-term information, expecting this can improve the performance of a simple RNN model.

Before doing the hyperparameter tuning, I standardized the data to speed up the convergence and reshaped them to fit with the input shape of RNN layers. For hyperparameter tuning, I set reasonable choice of intervals for the learning rate and number of neurons on each layer, as well as the choice of optimizers (Momentum optimization, Nesterov Accelerated Gradient, Adam, NAdam, and AdamW). On the first layer (LSTM), activation is set by ReLu to avoid vanishing gradient, and on the second layer, activation is set by LeakyReLu. For each RNN layer, I set dropout and recurrent dropout as parameters involved in the tuning process.

I did the hyperparameter tuning on data of Microsoft, and when applied to other stocks, I retrained them before doing the test and prediction. When applying the model to the index data, I re-do the hyperparameter tuning to find a model that can better fit with the Index’s characteristics. The code for hyperparameter tuning and training is modified from Lecture 17.


**SVM**

For the SVM regressor, my model would do hyperparameter training only on the top outperform stock “MSFT”, which is agreed among our group. And use this best parameter on the predictions of the other two stocks too.
I first standardize the data, since it is very crucial to the SVM regressor. Then, I use RandomizedSearchCV to help me find the best parameter combination of my model. The best parameter types I am searching are “C”, “gamma”, and “kernel”. I use this model to first predict the percentage price change of the three stocks separately for May 24, 2024. For each model, I keep the parameter unchanged, but retrain using each stock’s training data in order to better fit the new stock and do better predictions.
Second, I hyperparameter tune and train my model for the Index Average, and use the model with the best parameter to predict the percentage price change of the Index Average for May 24, 2024.
Finally, for each stock among the three stocks, I get the difference between the predicted data for Index and stock for May 24, 2024, and I also calculate the difference between the real data for Index and stock for that day. The variable “error” stands for the absolute value of the difference between the real difference and the predicted difference.


**Random Forest**

The final method that will be implemented is the Random Forest Regressor. This is a tree-based method, which could potentially provide a nice counterpoint to the other methods used in this project. Random forest is appropriate for this task, as was demonstrated in Lecture 5 and Homework 2. From these, we know that random forest models can perform decently well at time series prediction tasks. Since this method is tree-based, no scaling or normalization is necessary, and none is implemented.

We performed 3-fold cross validation and hyperparameter tuning over the hyperparameters maximum depth (max_depth) = 1, 5, 10, number of trees (n_estimators) = 10, 50, 100, 200, 400, complexity parameter for pruning (ccp_alpha) = 10e-1, 10e-3, 10e-5, 10e-7. Every combination of these hyperparameter values is tested to provide the most robust hyperparameter tuning. The code for this section is modified from Lecture 5.



# 2. Data Preprocessing and Feature Engineering*(Charlie Yin)*

All of our data are loaded from Yahoo Finance. Our original data are the daily stock prices and volume of Dow Jones Index and 30 stocks in the index. Our training set are from 2014-1-1 to 2020-1-1, validation set are from 2020-1-1 to 2023-1-1, our test set are from 2023-1-1 to 2024-5-9, and our prediction set are from 2024-5-9 to 2024-5-25. The prediction set are used to predict future price changes, meaning we want to predict the percentage of price change on May 24 (Friday).

In our study, we only consider close price and volume, and we then calculated the percentage of price change as a measure of performance, and percentage of volume as a measure of market sentiment and potential volatility.Considering different days of a week may also influence the performance of the stocks, we add a new features of days of a week by using OneHotEncoder Function in Python (By Sayako Mitchell) .

Our plan is to find out the top 3 tech stocks that outperform the Index for most times from our chosen time interval. Then train our model with the top 1 stock's data. Applying the best-performed model (in terms of RMSE) to predict the percentage change of price on May 24 (Friday) of all of the 3 top stocks and also the Index itself.

On Friday, we will firstly check if our prediction on the percentage change of price (for 3 stocks and the Index) are close to the real data, and then check if our prediction on the performance comparison between the top 3 stocks and the index is correct. And the results of these 2 steps will help us choose the best-performed model.

Here is the details of how did we find out the top 3 stocks: **The text and coding for this sub-part is completed by Tianrui Zhou**:
We want to find the top 3 tech stock, which means the top 3 tech stocks that outperform the index for the most times over our interval. For each stock in the 30 stocks, I do binomial encoding (1 and 0): if the closing price percentage change of that certain stock minus the closing price percentage change of the index is positive, then encode 1. Otherwise, encode 0. For each stock, the "outperform_score" is calculated by adding up all the "1"s in the array (the sum of the array). Likewise, I can calculate the outperform score for each of the 30 stocks. Through computation, we can see that the stocks 'MSFT', 'V' and 'AAPL' has the top 3 outperform score, which means they are the top three stocks that outperform the index the most among all the 30 stocks.


In [1]:
! pip install -q -U keras_tuner

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.1/129.1 kB 2.7 MB/s eta 0:00:00


In [2]:
! pip install yfinance

In [3]:
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
import keras_tuner as kt
import pandas as pd
import pandas_datareader as pdr
from datetime import datetime
import yfinance as yf
from sklearn.preprocessing import MinMaxScaler
from collections import defaultdict

In [4]:
Ticker_names = ["UNH","GS", "MSFT", "CAT", "HD", "AMGN", "CRM", "V", "MCD", "AXP",
                "TRV", "HON", "JPM", "AAPL", "AMZN", "BA", "IBM", "PG", "CVX", "JNJ",
                "MRK", "DIS", "MMM", "NKE", "WMT", "KO", "DOW", "CSCO", "VZ", "INTC"]

In [5]:
def data_generator(ticker_name):
    #Get data
    #newdf = df.reset_index() newdf
    Train_set = yf.download(ticker_name, start="2014-01-01", end="2020-01-01", interval = '1d', auto_adjust= False, progress=False)
    Valid_set = yf.download(ticker_name, start="2020-01-01", end="2023-01-01", interval = '1d', auto_adjust= False, progress=False)
    Test_set = yf.download(ticker_name, start="2023-01-01", end="2024-05-09", interval = '1d', auto_adjust= False, progress=False)
    prediction_set = yf.download(ticker_name, start="2024-05-09", end="2024-05-25", interval = '1d', auto_adjust= False, progress=False) # this set is used to predict future prices

    # add timestamp column
    Train_set['Timestamp'] = Train_set.index
    Valid_set['Timestamp'] = Valid_set.index
    Test_set['Timestamp'] = Test_set.index
    prediction_set['Timestamp'] = prediction_set.index

    #calculate percentages:
    train = Train_set[['Close', 'Volume']].pct_change() * 100
    valid = Valid_set[['Close', 'Volume']].pct_change() * 100
    test = Test_set[['Close', 'Volume']].pct_change() * 100
    predict = prediction_set[['Close', 'Volume']].pct_change() * 100

    # convert timestamp to day of the week
    Train_set['Timestamp'] = pd.to_datetime(Train_set['Timestamp'])
    Train_set['day_of_week'] = Train_set['Timestamp'].dt.dayofweek
    Valid_set['Timestamp'] = pd.to_datetime(Valid_set['Timestamp'])
    Valid_set['day_of_week'] = Valid_set['Timestamp'].dt.dayofweek
    Test_set['Timestamp'] = pd.to_datetime(Test_set['Timestamp'])
    Test_set['day_of_week'] = Test_set['Timestamp'].dt.dayofweek
    prediction_set['Timestamp'] = pd.to_datetime(prediction_set['Timestamp'])
    prediction_set['day_of_week'] = prediction_set['Timestamp'].dt.dayofweek

    # add day of week
    train['Day_of_Week']= Train_set['day_of_week']
    valid['Day_of_Week'] = Valid_set['day_of_week']
    test['Day_of_Week'] = Test_set['day_of_week']
    predict['Day_of_Week'] = prediction_set['day_of_week']

    # apply one-hot encoding for days of the week
    train_encoded = pd.get_dummies(train, columns=['Day_of_Week'])
    valid_encoded = pd.get_dummies(valid, columns=['Day_of_Week'])
    test_encoded = pd.get_dummies(test, columns=['Day_of_Week'])
    predict_encoded = pd.get_dummies(predict, columns=['Day_of_Week'])

    #rename the columns
    train_encoded.columns = ['Pct_Change_Price', 'Pct_Change_Volume', 'M','T','W','Th','F']
    valid_encoded.columns = ['Pct_Change_Price', 'Pct_Change_Volume','M','T','W','Th','F']
    test_encoded.columns = ['Pct_Change_Price', 'Pct_Change_Volume','M','T','W','Th','F']
    predict_encoded.columns = ['Pct_Change_Price', 'Pct_Change_Volume','M','T','W','Th','F']

    #drop NA
    train_data = train_encoded.dropna()
    valid_data = valid_encoded.dropna()
    test_data = test_encoded.dropna()
    predict_data = predict_encoded.dropna()

    #return the data frame
    return train_data, valid_data, test_data, predict_data

In [6]:
#Prepare Index data
data = yf.download("^DJI", start="2014-01-01", end="2024-05-25", interval = '1d', auto_adjust= False, progress=False)
df_pct_change = data[['Close', 'Volume']].pct_change() * 100
# Rename the columns appropriately
df_pct_change.columns = ['Pct_Change_Price', 'Pct_Change_Volume']
Index = df_pct_change.dropna()


In [7]:
def calculate_outperform_score(stock_data, index_data):
    '''
    Goal: Calculates the outperform score for a given stock data against the index data.

    Returns:
        The outperform score, which is the number of days the stock outperformed the index.
    '''
    outperform_array = (stock_data['Pct_Change_Price'] - index_data['Pct_Change_Price']) > 0
    outperform_score = outperform_array.sum()
    return outperform_score

outperform_scores = defaultdict(int)

In [8]:
for ticker in Ticker_names:
    train_data, valid_data, test_data, pred_data = data_generator(ticker)
    stock_data = pd.concat([train_data, valid_data, test_data, pred_data])

    # Calculate outperform score
    outperform_score = calculate_outperform_score(stock_data, Index)
    outperform_scores[ticker] = outperform_score

# Sort outperform scores and get top 3
sorted_scores = sorted(outperform_scores.items(), key=lambda x: x[1], reverse=True)

print("The top 3 stocks are:", sorted_scores[0:10])

The top 3 stocks are: [('MSFT', 1379), ('UNH', 1362), ('HD', 1357), ('V', 1355), ('TRV', 1351), ('AAPL', 1339), ('CSCO', 1332), ('CRM', 1329), ('AMZN', 1329), ('MCD', 1327)]


In [9]:
#Here is for data preprocessing
#X_train, y_train, X_valid, y_valid, X_test, y_test
def ts_split(ts, feature_steps=10, target_steps=1):
    n_obs = len(ts) - feature_steps - target_steps + 1
    X = np.array([ts[idx:idx + feature_steps] for idx in range(n_obs)])
    y = np.array([ts[idx + feature_steps:idx + feature_steps + target_steps] for idx in range(n_obs)])
    return X, y
def tsv_generator (data):
    pct_price = data["Pct_Change_Price"]
    pct_vol = data["Pct_Change_Volume"]
    M = data["M"]
    T = data['T']
    W = data["W"]
    Th = data['Th']
    F = data["F"]
    pct_price_X, pct_price_y = ts_split(pct_price, feature_steps=10, target_steps=1)
    pct_vol_X, pct_vol_y = ts_split(pct_vol, feature_steps=10, target_steps=1)
    M_X, M_y = ts_split(M, feature_steps=10, target_steps=1)
    T_X, T_y = ts_split(T, feature_steps=10, target_steps=1)
    W_X, W_y = ts_split(W, feature_steps=10, target_steps=1)
    Th_X, Th_y = ts_split(Th, feature_steps=10, target_steps=1)
    F_X, F_y = ts_split(F, feature_steps=10, target_steps=1)
    X = np.c_[pct_price_X,pct_vol_X, M_X, T_X, W_X, Th_X, F_X]
    y = pct_price_y
    return X, y

In [102]:
training_stock = "MSFT"
train, valid, test,predict  = data_generator(training_stock)
X_train, y_train = tsv_generator(train)
X_valid, y_valid = tsv_generator(valid)
X_test, y_test = tsv_generator(test)
X_predict, y_predict = tsv_generator(predict)

This is a sample of what our data looks like:

In [103]:
train

Pct_Change_Price  Pct_Change_Volume      M      T      W     Th  \
Date                                                                          
2014-01-03         -0.672766           1.640757  False  False  False  False   
2014-01-06         -2.113245          40.048113   True  False  False  False   
2014-01-07          0.774976         -17.890454  False   True  False  False   
2014-01-08         -1.785228          67.505614  False  False   True  False   
2014-01-09         -0.643175         -39.110781  False  False  False   True   
...                      ...                ...    ...    ...    ...    ...   
2019-12-24         -0.019058         -49.265727  False   True  False  False   
2019-12-26          0.819668          61.533841  False  False  False   True   
2019-12-27          0.182775          26.804677  False  False  False  False   
2019-12-30         -0.861859         -11.211766   True  False  False  False   
2019-12-31          0.069802          12.362066  False   True  False  False   

                F  
Date               
2014-01-03   True  
2014-01-06  False  
2014-01-07  False  
2014-01-08  False  
2014-01-09  False  
...           ...  
2019-12-24  False  
2019-12-26  False  
2019-12-27   True  
2019-12-30  False  
2019-12-31  False  

[1509 rows x 7 columns]

# 3. Recurrent Neural Network Model *Charlie Yin*

In [101]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_valid = scaler.transform(X_valid)
X_test = scaler.transform(X_test)
X_predict = scaler.transform(X_predict)

In [13]:
X_train = np.reshape(X_train, (X_train.shape[0], X_train.shape[1],1))
y_train = np.reshape(y_train, (y_train.shape[0],1))

X_valid = np.reshape(X_valid, (X_valid.shape[0], X_valid.shape[1],1))
y_valid = np.reshape(y_valid, (y_valid.shape[0],1))

X_test = np.reshape(X_test, (X_test.shape[0], X_test.shape[1],1))
y_test = np.reshape(y_test, (y_test.shape[0],1))

X_predict = np.reshape(X_predict, (X_predict.shape[0], X_predict.shape[1],1))
y_predict = np.reshape(y_predict, (y_predict.shape[0],1))

In [14]:
def reset_session(seed=42):
    tf.random.set_seed(seed)
    np.random.seed(seed)
    tf.keras.backend.clear_session()

Hyperparameter tuning for RNN model

In [36]:
def build_RNN_model(hp):
    #layers
    n_hidden = hp.Int("n_hidden", min_value=1, max_value=10)
    #neurons
    n_neurons = hp.Int("n_neurons", min_value=1, max_value=150)
    #learning_rate
    learning_rate = hp.Float("learning_rate", min_value=1e-4, max_value=1,
                             sampling="log")
    #optimizer
    optimizer = hp.Choice("optimizer", values=["momentum optimization","NAG", "adam", "nadam", "AdamW"])
    momentum = hp.Float("momentum", min_value=0.0, max_value = 1.0)
    weight_decay = hp.Float("weight_decay", min_value = 1e-6, max_value = 1)

    dropout = hp.Float("dropout", min_value = 0, max_value = 0.9)
    recurrent_dropout = hp.Float("recurrent_dropout", min_value = 0, max_value = 0.9)

    if optimizer == "momentum optimization":
        optimizer = tf.keras.optimizers.SGD(learning_rate=learning_rate, momentum=momentum)
    elif optimizer == "NAG":
        optimizer = tf.keras.optimizers.SGD(learning_rate=learning_rate, momentum=momentum,
                                    nesterov=True)
    elif optimizer == "adam":
        optimizer = tf.keras.optimizers.Adam(learning_rate=learning_rate)
    elif optimizer == "nadam":
        optimizer = tf.keras.optimizers.Nadam(learning_rate=learning_rate)

    else: #adamw
        optimizer = tf.keras.optimizers.AdamW(weight_decay = weight_decay, learning_rate=learning_rate)

    model = tf.keras.Sequential()

    lstm_units = hp.Int("lstm_units", min_value=1, max_value=150)
    model.add(tf.keras.layers.LSTM(lstm_units, return_sequences=True, activation="relu", input_shape=(70, 1)))

    # Add RNN layers based on n_hidden
    for i in range(n_hidden):
        model.add(tf.keras.layers.SimpleRNN(n_neurons, return_sequences=True, dropout=dropout,
                                             recurrent_dropout=recurrent_dropout, activation="leaky_relu"))

    # Final RNN layer
    model.add(tf.keras.layers.SimpleRNN(n_neurons, dropout=dropout, recurrent_dropout=recurrent_dropout))

    model.add(tf.keras.layers.Dense(1))  # Predicting the next step's price

    model.compile(loss="mse", optimizer=optimizer)

    return model


In [16]:
random_search_tuner = kt.RandomSearch(
    build_RNN_model, objective="val_loss", max_trials=10,  overwrite=True)

random_search_tuner.search(X_train[:1000], y_train[:1000], epochs=20,
                           validation_data=(X_valid, y_valid)
                          )

best_trial = random_search_tuner.oracle.get_best_trials(num_trials=1)[0]
best_trial.summary()

Trial 10 Complete [00h 01m 27s]
val_loss: 4.848118782043457

Best val_loss So Far: 4.837200164794922
Total elapsed time: 00h 29m 15s
Trial 01 summary
Hyperparameters:
n_hidden: 6
n_neurons: 42
learning_rate: 0.002264838487129011
optimizer: nadam
momentum: 0.11620353199188593
weight_decay: 0.33468706641372403
dropout: 0.7609906851645137
recurrent_dropout: 0.6193845052764336
lstm_units: 57
Score: 4.837200164794922


Train and test the RNN model

In [17]:
best_models = random_search_tuner.get_best_models(num_models=1)
best_model = best_models[0]

early_stopping_cb = tf.keras.callbacks.EarlyStopping(patience=10,
                                                  mode="max")

run_model = best_model.fit(X_train, y_train, epochs=100, batch_size=32,
               validation_data=(X_valid, y_valid),
               callbacks=[early_stopping_cb])

best_model.evaluate(X_test, y_test)

Epoch 1/100
47/47 [==============================] - 24s 239ms/step - loss: 2.0617 - val_loss: 4.8374
Epoch 2/100
47/47 [==============================] - 10s 218ms/step - loss: 2.0327 - val_loss: 4.8377
Epoch 3/100
47/47 [==============================] - 10s 223ms/step - loss: 2.0481 - val_loss: 4.8385
Epoch 4/100
47/47 [==============================] - 8s 181ms/step - loss: 2.0387 - val_loss: 4.8384
Epoch 5/100
47/47 [==============================] - 10s 217ms/step - loss: 2.0367 - val_loss: 4.8384
Epoch 6/100
47/47 [==============================] - 10s 211ms/step - loss: 2.0458 - val_loss: 4.8408
Epoch 7/100
47/47 [==============================] - 9s 194ms/step - loss: 2.0438 - val_loss: 4.8399
Epoch 8/100
47/47 [==============================] - 10s 205ms/step - loss: 2.0454 - val_loss: 4.8519
Epoch 9/100
47/47 [==============================] - 10s 212ms/step - loss: 2.0368 - val_loss: 4.8706
Epoch 10/100
47/47 [==============================] - 10s 213ms/step - loss: 2.0388 

2.199357509613037

Test Loss of RNN Model-MSE

In [18]:
test_loss = best_model.evaluate(X_test, y_test)
np.sqrt(test_loss)

11/11 [==============================] - 0s 31ms/step - loss: 2.1994


1.4830230981387436

Predict Next MSFT price change

In [19]:
predict_MSFT = best_model.predict(X_predict)
predict_MSFT

1/1 [==============================] - 1s 1s/step


array([[0.0983042]], dtype=float32)

In [20]:
y_predict

array([[0.7400477]])

Real:0.74

Predict Next V price change

In [21]:
training_stock = "V"
train, valid, test,predict  = data_generator(training_stock)
X_train, y_train = tsv_generator(train)
X_valid, y_valid = tsv_generator(valid)
X_test, y_test = tsv_generator(test)
X_predict, y_predict = tsv_generator(predict)

In [22]:
X_train = np.reshape(X_train, (X_train.shape[0], X_train.shape[1],1))
y_train = np.reshape(y_train, (y_train.shape[0],1))

X_valid = np.reshape(X_valid, (X_valid.shape[0], X_valid.shape[1],1))
y_valid = np.reshape(y_valid, (y_valid.shape[0],1))

X_test = np.reshape(X_test, (X_test.shape[0], X_test.shape[1],1))
y_test = np.reshape(y_test, (y_test.shape[0],1))

X_predict = np.reshape(X_predict, (X_predict.shape[0], X_predict.shape[1],1))
y_predict_V = np.reshape(y_predict, (y_predict.shape[0],1))

In [23]:
reset_session()
best_models = random_search_tuner.get_best_models(num_models=1)
best_model = best_models[0]

early_stopping_cb = tf.keras.callbacks.EarlyStopping(patience=10,
                                                  mode="max")

run_model = best_model.fit(X_train, y_train, epochs=100, batch_size=32,
               validation_data=(X_valid, y_valid),
               callbacks=[early_stopping_cb])

predict_V = best_model.predict(X_predict)
predict_V

Epoch 1/100
47/47 [==============================] - 24s 230ms/step - loss: 1.7000 - val_loss: 4.4946
Epoch 2/100
47/47 [==============================] - 8s 179ms/step - loss: 1.6993 - val_loss: 4.4931
Epoch 3/100
47/47 [==============================] - 10s 215ms/step - loss: 1.6865 - val_loss: 4.4928
Epoch 4/100
47/47 [==============================] - 10s 216ms/step - loss: 1.7045 - val_loss: 4.4941
Epoch 5/100
47/47 [==============================] - 8s 179ms/step - loss: 1.6888 - val_loss: 4.5128
Epoch 6/100
47/47 [==============================] - 10s 216ms/step - loss: 1.6943 - val_loss: 4.5052
Epoch 7/100
47/47 [==============================] - 10s 215ms/step - loss: 1.6891 - val_loss: 4.5022
Epoch 8/100
47/47 [==============================] - 10s 223ms/step - loss: 1.6966 - val_loss: 4.5080
Epoch 9/100
47/47 [==============================] - 8s 180ms/step - loss: 1.6867 - val_loss: 4.5174
Epoch 10/100
47/47 [==============================] - 10s 215ms/step - loss: 1.6866 -

array([[0.10754932]], dtype=float32)

In [24]:
y_predict_V

array([[0.09480335]])

Predict Next AAPL price change

In [25]:
training_stock = "AAPL"
train, valid, test,predict  = data_generator(training_stock)
X_train, y_train = tsv_generator(train)
X_valid, y_valid = tsv_generator(valid)
X_test, y_test = tsv_generator(test)
X_predict, y_predict = tsv_generator(predict)

In [26]:
X_train = np.reshape(X_train, (X_train.shape[0], X_train.shape[1],1))
y_train = np.reshape(y_train, (y_train.shape[0],1))

X_valid = np.reshape(X_valid, (X_valid.shape[0], X_valid.shape[1],1))
y_valid = np.reshape(y_valid, (y_valid.shape[0],1))

X_test = np.reshape(X_test, (X_test.shape[0], X_test.shape[1],1))
y_test = np.reshape(y_test, (y_test.shape[0],1))

X_predict = np.reshape(X_predict, (X_predict.shape[0], X_predict.shape[1],1))
y_predict_AAPL = np.reshape(y_predict, (y_predict.shape[0],1))

In [27]:
reset_session()
best_models = random_search_tuner.get_best_models(num_models=1)
best_model = best_models[0]

early_stopping_cb = tf.keras.callbacks.EarlyStopping(patience=10,
                                                  mode="max")

run_model = best_model.fit(X_train, y_train, epochs=100, batch_size=32,
               validation_data=(X_valid, y_valid),
               callbacks=[early_stopping_cb])

predict_AAPL = best_model.predict(X_predict)
predict_AAPL

Epoch 1/100
47/47 [==============================] - 24s 229ms/step - loss: 2.3674 - val_loss: 5.4570
Epoch 2/100
47/47 [==============================] - 9s 182ms/step - loss: 2.3779 - val_loss: 5.4574
Epoch 3/100
47/47 [==============================] - 10s 217ms/step - loss: 2.3681 - val_loss: 5.4580
Epoch 4/100
47/47 [==============================] - 10s 217ms/step - loss: 2.3790 - val_loss: 5.4579
Epoch 5/100
47/47 [==============================] - 8s 179ms/step - loss: 2.3553 - val_loss: 5.4585
Epoch 6/100
47/47 [==============================] - 10s 216ms/step - loss: 2.3776 - val_loss: 5.4570
Epoch 7/100
47/47 [==============================] - 10s 216ms/step - loss: 2.3542 - val_loss: 5.4586
Epoch 8/100
47/47 [==============================] - 9s 192ms/step - loss: 2.3977 - val_loss: 5.4596
Epoch 9/100
47/47 [==============================] - 10s 204ms/step - loss: 2.3669 - val_loss: 5.4622
Epoch 10/100
47/47 [==============================] - 10s 217ms/step - loss: 2.3694 -

array([[0.11547567]], dtype=float32)

In [28]:
y_predict_AAPL

array([[1.65881355]])

Predict next DJI price

In [48]:
training_stock = "^DJI"
train, valid, test,predict  = data_generator(training_stock)
X_train, y_train = tsv_generator(train)
X_valid, y_valid = tsv_generator(valid)
X_test, y_test = tsv_generator(test)
X_predict, y_predict = tsv_generator(predict)

In [49]:
from sklearn.preprocessing import StandardScaler

scaler_1 = StandardScaler()

X_train = scaler_1.fit_transform(X_train)
X_valid = scaler_1.fit_transform(X_valid)
X_test = scaler_1.fit_transform(X_test)
X_predict = scaler_1.fit_transform(X_predict)

In [50]:
X_train = np.reshape(X_train, (X_train.shape[0], X_train.shape[1],1))
y_train = np.reshape(y_train, (y_train.shape[0],1))

X_valid = np.reshape(X_valid, (X_valid.shape[0], X_valid.shape[1],1))
y_valid = np.reshape(y_valid, (y_valid.shape[0],1))

X_test = np.reshape(X_test, (X_test.shape[0], X_test.shape[1],1))
y_test = np.reshape(y_test, (y_test.shape[0],1))

X_predict = np.reshape(X_predict, (X_predict.shape[0], X_predict.shape[1],1))
y_predict_DJI = np.reshape(y_predict, (y_predict.shape[0],1))

In [51]:
random_search_tuner = kt.RandomSearch(
    build_RNN_model, objective="val_loss", max_trials=5,  overwrite=True)

random_search_tuner.search(X_train[:1000], y_train[:1000], epochs=20,
                           validation_data=(X_valid, y_valid)
                          )

best_trial = random_search_tuner.oracle.get_best_trials(num_trials=1)[0]
best_trial.summary()

Trial 5 Complete [00h 04m 42s]
val_loss: nan

Best val_loss So Far: 2.5398988723754883
Total elapsed time: 00h 17m 53s
Trial 0 summary
Hyperparameters:
n_hidden: 8
n_neurons: 34
learning_rate: 0.019023652230789233
optimizer: adam
momentum: 0.1397735289488119
weight_decay: 0.9149576743714959
dropout: 0.7998929775629741
recurrent_dropout: 0.8386385892758781
lstm_units: 27
Score: 2.5398988723754883


In [52]:
best_models = random_search_tuner.get_best_models(num_models=1)
best_model = best_models[0]

early_stopping_cb = tf.keras.callbacks.EarlyStopping(patience=10,
                                                  mode="max")

run_model = best_model.fit(X_train, y_train, epochs=100, batch_size=32,
               validation_data=(X_valid, y_valid),
               callbacks=[early_stopping_cb])

best_model.evaluate(X_test, y_test)

Epoch 1/100
47/47 [==============================] - 26s 272ms/step - loss: 0.7276 - val_loss: 2.5402
Epoch 2/100
47/47 [==============================] - 11s 245ms/step - loss: 0.7355 - val_loss: 2.5757
Epoch 3/100
47/47 [==============================] - 11s 244ms/step - loss: 0.7379 - val_loss: 2.7171
Epoch 4/100
47/47 [==============================] - 11s 243ms/step - loss: 0.7309 - val_loss: 2.5400
Epoch 5/100
47/47 [==============================] - 12s 250ms/step - loss: 0.7465 - val_loss: 2.5419
Epoch 6/100
47/47 [==============================] - 10s 209ms/step - loss: 0.7381 - val_loss: 2.5401
Epoch 7/100
47/47 [==============================] - 11s 243ms/step - loss: 0.7389 - val_loss: 2.5856
Epoch 8/100
47/47 [==============================] - 11s 244ms/step - loss: 0.7433 - val_loss: 2.5778
Epoch 9/100
47/47 [==============================] - 11s 243ms/step - loss: 0.7526 - val_loss: 2.5431
Epoch 10/100
47/47 [==============================] - 12s 251ms/step - loss: 0.735

0.48032987117767334

In [53]:
reset_session()
best_models = random_search_tuner.get_best_models(num_models=1)
best_model = best_models[0]

early_stopping_cb = tf.keras.callbacks.EarlyStopping(patience=10,
                                                  mode="max")

run_model = best_model.fit(X_train, y_train, epochs=100, batch_size=32,
               validation_data=(X_valid, y_valid),
               callbacks=[early_stopping_cb])

predict_DJI = best_model.predict(X_predict)
predict_DJI

Epoch 1/100
47/47 [==============================] - 27s 288ms/step - loss: 0.7290 - val_loss: 2.5557
Epoch 2/100
47/47 [==============================] - 10s 222ms/step - loss: 0.7386 - val_loss: 2.5407
Epoch 3/100
47/47 [==============================] - 11s 232ms/step - loss: 0.7548 - val_loss: 2.5964
Epoch 4/100
47/47 [==============================] - 11s 245ms/step - loss: 0.7438 - val_loss: 2.5398
Epoch 5/100
47/47 [==============================] - 11s 246ms/step - loss: 0.7241 - val_loss: 2.6093
Epoch 6/100
47/47 [==============================] - 11s 245ms/step - loss: 0.7295 - val_loss: 2.5514
Epoch 7/100
47/47 [==============================] - 12s 255ms/step - loss: 0.7251 - val_loss: 2.6079
Epoch 8/100
47/47 [==============================] - 10s 220ms/step - loss: 0.7314 - val_loss: 2.5514
Epoch 9/100
47/47 [==============================] - 11s 233ms/step - loss: 0.7317 - val_loss: 2.6337
Epoch 10/100
47/47 [==============================] - 12s 247ms/step - loss: 0.732

array([[0.26170298]], dtype=float32)

In [54]:
y_predict_DJI

array([[0.01107922]])

Comparison

In [104]:
predict_diff = predict_DJI - predict_MSFT
real_diff = y_predict_DJI  - y_predict
error = np.abs(real_diff-predict_diff)
print("The predict difference between the Index and the stock price change is", predict_diff)
print("The real difference between the Index and the stock price change is", real_diff)


The predict difference between the Index and the stock price change is [[0.16339877]]
The real difference between the Index and the stock price change is [[-0.72896848]]
The absolute value of the distance between real and predict is [[0.89236725]]


In [56]:
predict_diff = predict_DJI - predict_V
real_diff = y_predict_DJI  - y_predict_V
error = np.abs(real_diff-predict_diff)
print("The predict difference between the Index and the stock price change is", predict_diff)
print("The real difference between the Index and the stock price change is", real_diff)
print("The absolute value of the distance between real and predict is", error)

The predict difference between the Index and the stock price change is [[0.15415367]]
The real difference between the Index and the stock price change is [[-0.08372414]]
The absolute value of the distance between real and predict is [[0.23787781]]


In [57]:
predict_diff = predict_DJI - predict_AAPL
real_diff = y_predict_DJI  - y_predict_AAPL
error = np.abs(real_diff-predict_diff)
print("The predict difference between the Index and the stock price change is", predict_diff)
print("The real difference between the Index and the stock price change is", real_diff)
print("The absolute value of the distance between real and predict is", error)

The predict difference between the Index and the stock price change is [[0.14622732]]
The real difference between the Index and the stock price change is [[-1.64773433]]
The absolute value of the distance between real and predict is [[1.79396165]]


# 4. SVM Model *(Tianrui Zhou)*

In [58]:
# Use the top stock to find the best parameter for the model
training_stock = "MSFT"
train, valid, test, predict = data_generator(training_stock)
X_train, y_train = tsv_generator(train)
X_valid, y_valid = tsv_generator(valid)
X_test, y_test = tsv_generator(test)
X_predict, y_predict = tsv_generator(predict)

In [59]:
train

Pct_Change_Price  Pct_Change_Volume      M      T      W     Th  \
Date                                                                          
2014-01-03         -0.672766           1.640757  False  False  False  False   
2014-01-06         -2.113245          40.048113   True  False  False  False   
2014-01-07          0.774976         -17.890454  False   True  False  False   
2014-01-08         -1.785228          67.505614  False  False   True  False   
2014-01-09         -0.643175         -39.110781  False  False  False   True   
...                      ...                ...    ...    ...    ...    ...   
2019-12-24         -0.019058         -49.265727  False   True  False  False   
2019-12-26          0.819668          61.533841  False  False  False   True   
2019-12-27          0.182775          26.804677  False  False  False  False   
2019-12-30         -0.861859         -11.211766   True  False  False  False   
2019-12-31          0.069802          12.362066  False   True  False  False   

                F  
Date               
2014-01-03   True  
2014-01-06  False  
2014-01-07  False  
2014-01-08  False  
2014-01-09  False  
...           ...  
2019-12-24  False  
2019-12-26  False  
2019-12-27   True  
2019-12-30  False  
2019-12-31  False  

[1509 rows x 7 columns]

In [60]:
X_train

array([[-0.67276642, -2.11324514,  0.77497584, ...,  0.        ,
         0.        ,  0.        ],
       [-2.11324514,  0.77497584, -1.78522804, ...,  0.        ,
         0.        ,  1.        ],
       [ 0.77497584, -1.78522804, -0.64317548, ...,  0.        ,
         1.        ,  0.        ],
       ...,
       [ 1.01516716,  0.84181235,  0.64712354, ...,  0.        ,
         0.        ,  0.        ],
       [ 0.84181235,  0.64712354, -0.54008638, ...,  0.        ,
         0.        ,  1.        ],
       [ 0.64712354, -0.54008638, -0.20687008, ...,  0.        ,
         1.        ,  0.        ]])

In [61]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_valid = scaler.transform(X_valid)
X_test = scaler.transform(X_test)
X_predict = scaler.transform(X_predict)

In [62]:
from sklearn.metrics import mean_squared_error
from sklearn.svm import SVR
from sklearn.model_selection import RandomizedSearchCV

# Use random search to search for the best parameter for the model

svr = SVR()

param_distributions = {
    'C': [0.001, 0.01, 0.1, 1, 10],
    'gamma': ['scale', 'auto', 0.01, 0.1, 0.5, 1],
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid']
}

random_search = RandomizedSearchCV(
    estimator = svr,
    param_distributions = param_distributions,
    n_iter = 20,
    scoring='neg_root_mean_squared_error',
    cv=3,
    random_state=42,
    n_jobs=-1
)

random_search.fit(X_train, y_train)

# Best model
rmse = -random_search.best_score_
print("The RMSE of the model with the best parameter is",rmse)

/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/fork_exec.py:38: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  pid = os.fork()


The RMSE of the model with the best parameter is 1.4159382715000444


/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [63]:
best_model = random_search.best_estimator_
print("The best model is",best_model)

The best model is SVR(C=0.001, gamma='auto', kernel='linear')


In [64]:
y_val_pred = best_model.predict(X_valid)

# Calculate RMSE
rmse_val = mean_squared_error(y_valid, y_val_pred, squared=False)
print("The RMSE on validation set is", rmse_val)

The RMSE on validation set is 2.1865756928239213


In [65]:
y_test_pred = best_model.predict(X_test)

# Calculate RMSE
rmse_test = mean_squared_error(y_test, y_test_pred, squared=False)
print("The RMSE on test set is", rmse_test)

The RMSE on test set is 1.476589283222646


In [66]:
X_predict[-1]

array([ 0.33614762, -0.24577747,  0.40506648,  1.0194197 , -0.42186091,
       -0.20623613,  0.77829549,  0.53174712,  0.16500191, -0.64925601,
       -0.34995599,  0.22035702, -0.19178858,  0.98214273, -0.64495686,
       -0.43623542,  0.00214118,  0.61465805, -0.51509613, -0.2541934 ,
       -0.48137027,  2.07740291, -0.48031872, -0.48031872, -0.48031872,
       -0.48137027,  2.07740291, -0.48031872, -0.48031872, -0.48137027,
       -0.50749417, -0.50853339,  1.9664392 , -0.50749417, -0.50749417,
       -0.50749417, -0.50853339,  1.9664392 , -0.50749417, -0.50749417,
       -0.50853339, -0.50853339, -0.50957224,  1.9624303 , -0.50853339,
       -0.50853339, -0.50853339, -0.50853339,  1.9664392 , -0.50749417,
       -0.50125052, -0.50125052, -0.50125052, -0.50229216,  1.99087322,
       -0.50125052, -0.50125052, -0.50229216, -0.50229216,  1.99087322,
        1.99501039, -0.50020846, -0.50020846, -0.50020846, -0.50125052,
        1.99501039, -0.50020846, -0.50020846, -0.50125052, -0.50

**Stock MSFT Prediction**

In [67]:
X_last10 = X_predict[-1:]
y_last10_pred = best_model.predict(X_last10)
StockMSFT_pred = y_last10_pred[0]
print("For stock "+training_stock+", the predict percent of price change on 2024.05.24 is", StockMSFT_pred)

For stock MSFT, the predict percent of price change on 2024.05.24 is 0.11301573436796555


In [68]:
StockMSFT_real = y_predict[-1]
print("For stock "+training_stock+", the real percent of price change on 2024.05.24 is", StockMSFT_real)

For stock MSFT, the real percent of price change on 2024.05.24 is [0.7400477]


**Stock V Prediction**

In [69]:
current_stock = "V"
train, valid, test, predict = data_generator(current_stock)
X_train, y_train = tsv_generator(train)
X_predict, y_predict = tsv_generator(predict)

X_train = scaler.fit_transform(X_train)
X_predict = scaler.transform(X_predict)

best_model.fit(X_train, y_train)

X_last10 = X_predict[-1:]
y_last10_pred = best_model.predict(X_last10)

StockV_pred = y_last10_pred[0]
print("For stock "+current_stock+", the predict percent of price change on 2024.05.24 is", StockV_pred)

For stock V, the predict percent of price change on 2024.05.24 is 0.19722774329980433


/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [70]:
StockV_real = y_predict[-1]
print("For stock "+current_stock+", the real percent of price change on 2024.05.24 is", StockV_real)

For stock V, the real percent of price change on 2024.05.24 is [0.09480335]


**Stock AAPL Prediction**

In [71]:
current_stock = "AAPL"
train, valid, test, predict = data_generator(current_stock)
X_train, y_train = tsv_generator(train)
X_predict, y_predict = tsv_generator(predict)

X_train = scaler.fit_transform(X_train)
X_predict = scaler.transform(X_predict)

best_model.fit(X_train, y_train)

X_last10 = X_predict[-1:]
y_last10_pred = best_model.predict(X_last10)

StockAAPL_pred = y_last10_pred[0]
print("For stock "+current_stock+", the predict percent of price change on 2024.05.24 is", StockAAPL_pred)

For stock AAPL, the predict percent of price change on 2024.05.24 is 0.07373272360948092


/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [72]:
StockAAPL_real = y_predict[-1]
print("For stock "+current_stock+", the real percent of price change on 2024.05.24 is", StockAAPL_real)

For stock AAPL, the real percent of price change on 2024.05.24 is [1.65881355]


**Now, train and hypertuning for te Index**

In [73]:
train, valid, test, predict = data_generator("^DJI")
X_train, y_train_index = tsv_generator(train)
X_valid, y_valid_index = tsv_generator(valid)
X_test, y_test_index = tsv_generator(test)
X_predict, y_predict_index = tsv_generator(predict)

In [74]:
scaler = StandardScaler()

X_train_index = scaler.fit_transform(X_train)
X_valid_index = scaler.transform(X_valid)
X_test_index = scaler.transform(X_test)
X_predict_index = scaler.transform(X_predict)

In [75]:
svr = SVR()

param_distributions = {
    'C': [0.0001, 0.001, 0.01, 0.1, 1, 10],
    'gamma': ['scale', 'auto', 0.01, 0.1, 0.3, 0.5, 0.7, 1, 10, 20, 50, 100],
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid']
}

random_search = RandomizedSearchCV(
    estimator = svr,
    param_distributions = param_distributions,
    n_iter = 20,
    scoring='neg_root_mean_squared_error',
    cv=3,
    random_state=42,
    n_jobs=-1
)

random_search.fit(X_train_index, y_train_index)

# Best model
rmse = -random_search.best_score_
print("The RMSE of the model with the best parameter is",rmse)

The RMSE of the model with the best parameter is 0.8157626568959953


/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [76]:
best_model_Index = random_search.best_estimator_
print("The best model for the Index is",best_model_Index)

The best model for the Index is SVR(C=0.0001, gamma=0.01, kernel='poly')


In [77]:
y_val_pred = best_model_Index.predict(X_valid_index)

# Calculate RMSE
rmse_val = mean_squared_error(y_valid_index, y_val_pred, squared=False)
print("For Index, the RMSE on validation set is", rmse_val)

For Index, the RMSE on validation set is 1.5938952312527201


In [78]:
y_test_pred = best_model_Index.predict(X_test_index)

# Calculate RMSE
rmse_test = mean_squared_error(y_test_index, y_test_pred, squared=False)
print("For Index, the RMSE on test set is", rmse_test)

For Index, the RMSE on test set is 0.674280776099789


In [79]:
#Do the prediction for Index

X_last10_index = X_predict_index[-1:]
y_last10_pred_index = best_model_Index.predict(X_last10_index)
Index_pred = y_last10_pred_index[0]
print("The predict percent of price change of Index on 2024.05.28 is", Index_pred)

The predict percent of price change of Index on 2024.05.28 is 0.05793381285985584


In [80]:
Index_real = y_predict_index[-1]
print("The predict percent of price change of Index on 2024.05.28 is", Index_real)

The predict percent of price change of Index on 2024.05.28 is [0.01107922]


**For each of the three stocks, compare the prediction difference and real difference between the stock and Index:**

* Stock MSFT

In [81]:
predict_diff = Index_pred - StockMSFT_pred
real_diff = Index_real - StockMSFT_real
error = np.abs(real_diff-predict_diff)
print("The predict difference between the Index and the stock price change is", predict_diff)
print("The real difference between the Index and the stock price change is", real_diff)
print("The absolute value of the distance between real and predict is", error)

The predict difference between the Index and the stock price change is -0.05508192150810971
The real difference between the Index and the stock price change is [-0.72896848]
The absolute value of the distance between real and predict is [0.67388656]


* Stock V

In [82]:
predict_diff = Index_pred - StockV_pred
real_diff = Index_real - StockV_real
error = np.abs(real_diff-predict_diff)
print("The predict difference between the Index and the stock price change is", predict_diff)
print("The real difference between the Index and the stock price change is", real_diff)
print("The absolute value of the distance between real and predict is", error)

The predict difference between the Index and the stock price change is -0.13929393043994848
The real difference between the Index and the stock price change is [-0.08372414]
The absolute value of the distance between real and predict is [0.05556979]


* Stock AAPL

In [83]:
predict_diff = Index_pred - StockAAPL_pred
real_diff = Index_real - StockAAPL_real
error = np.abs(real_diff-predict_diff)
print("The predict difference between the Index and the stock price change is", predict_diff)
print("The real difference between the Index and the stock price change is", real_diff)
print("The absolute value of the distance between real and predict is", error)

The predict difference between the Index and the stock price change is -0.015798910749625084
The real difference between the Index and the stock price change is [-1.64773433]
The absolute value of the distance between real and predict is [1.63193542]


# 5. Random Forest Model *(Sayako Mitchell)*

The model I chose to investigate is Random Forest Regressor. This model was chosen because it was shown to have decently good performance on time series tasks, as demonstrated in Lecture 5 and Homework 2. [INSERT ALGORITHM]

In [84]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import TimeSeriesSplit
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt

train, valid, test, predict = data_generator("MSFT")
X_train, y_train = tsv_generator(train)
X_valid, y_valid = tsv_generator(valid)
X_test, y_test = tsv_generator(test)
X_predict, y_predict = tsv_generator(predict)

tscv = TimeSeriesSplit(n_splits=3)

# Hyperparamter combinations
max_depth_list = [1, 5, 10]
n_estimators_list = [10, 50, 100, 200, 400]
k = [-1, -3, -5, -7]
alpha_list = [10**ks for ks in k]

series_len = y_train.size

def time_series_valid_test(X, y, n_split, valid_or_test, optimal_par=None):
    tscv = TimeSeriesSplit(n_splits=n_split)
    rf_rmse = []
    i = 0
    for train_index, test_index in tscv.split(X):
        # Tuning
        if valid_or_test == "valid":
            for max_depth in max_depth_list:
              for alpha in alpha_list:
                for n_estimators in n_estimators_list:
                    model_rf = RandomForestRegressor(random_state=42,
                               max_depth=max_depth, ccp_alpha=alpha,
                               n_estimators=n_estimators, n_jobs=-1)
                    model_rf.fit(X_train, y_train.ravel())
                    y_val_rf = model_rf.predict(X_valid)
                    rf_rmse.append(np.sqrt(mean_squared_error(y_valid, y_val_rf)))

        # Evalulate on test set
        if valid_or_test == "test":
            model_rf = RandomForestRegressor(random_state=42,
                       max_depth=optimal_par[0], ccp_alpha=optimal_par[1], n_estimators=optimal_par[2])
            model_rf.fit(X_train, y_train.ravel())
            y_test_rf = model_rf.predict(X_test)
            rf_rmse.append(np.sqrt(mean_squared_error(y_test, y_test_rf)))

            # Plot the prediction for the last CV fold
            if i == n_split:
                plt.plot(range(series_len-test_index.size,series_len),
                         y_test_rf, label="2-steps ahead prediction")
                plt.plot(range(series_len-test_index.size,series_len),
                         y_test, label="True value")
                plt.legend(loc="upper left")
                plt.show()

    # Average RMSE over CV folds
    if valid_or_test == "valid":
        rf_rmse = np.mean(np.array(rf_rmse).reshape(
            n_split, len(max_depth_list)*len(n_estimators_list)*len(alpha_list)), axis=0)
        return rf_rmse
    if valid_or_test == "test":
        rf_rmse = np.mean(rf_rmse)
        return rf_rmse, y_test_rf

**Hyperparameter tuning**

In [85]:
rf_rmse = time_series_valid_test(X_train, y_train, 3, "valid")
ind = 0
min_rmse = 10
for max_depth in max_depth_list:
  for alpha in alpha_list:
        for n_estimators in n_estimators_list:
            if rf_rmse[ind] < min_rmse:
                min_rmse = rf_rmse[ind]
                best_md = max_depth
                best_alpha = alpha
                best_ne = n_estimators
            print(["(max_depth, alpha, n_estimators):",[max_depth, alpha, n_estimators]])
            print(rf_rmse[ind])
            ind += 1

['(max_depth, alpha, n_estimators):', [1, 0.1, 10]]
2.200358658812641
['(max_depth, alpha, n_estimators):', [1, 0.1, 50]]
2.1980098511659287
['(max_depth, alpha, n_estimators):', [1, 0.1, 100]]
2.198167246974628
['(max_depth, alpha, n_estimators):', [1, 0.1, 200]]
2.1988665903493554
['(max_depth, alpha, n_estimators):', [1, 0.1, 400]]
2.1989966866652786
['(max_depth, alpha, n_estimators):', [1, 0.001, 10]]
2.190559480718384
['(max_depth, alpha, n_estimators):', [1, 0.001, 50]]
2.1869944548764124
['(max_depth, alpha, n_estimators):', [1, 0.001, 100]]
2.186583855338965
['(max_depth, alpha, n_estimators):', [1, 0.001, 200]]
2.1862203442415766
['(max_depth, alpha, n_estimators):', [1, 0.001, 400]]
2.186543953907603
['(max_depth, alpha, n_estimators):', [1, 1e-05, 10]]
2.190559480718384
['(max_depth, alpha, n_estimators):', [1, 1e-05, 50]]
2.1869944548764124
['(max_depth, alpha, n_estimators):', [1, 1e-05, 100]]
2.186583855338965
['(max_depth, alpha, n_estimators):', [1, 1e-05, 200]]
2.1862

In [86]:
# validation rmse
# hyperparameters:
print(min_rmse, best_md, best_alpha, best_ne)
min(rf_rmse)

2.1735946589068007 5 1e-07 400


2.1735946589068007

In [87]:
rf_rmse, y_test_rf = time_series_valid_test(X_test, y_test, 3, "test", [best_md, best_alpha, best_ne])
rf_rmse

1.489768498761676

After hyperparameter tuning, we see that the optimal parameters for this model are maximum depth = 5, alpha = 1e-7, and n_estimators = 400. This results in a validation RMSE of 2.1736 and a test RMSE of 1.4898.

**Predicting the percentage change of MSFT, AAPL, V**

In [88]:
# MSFT
best_rf = RandomForestRegressor(random_state=42, max_depth=best_md, ccp_alpha=best_alpha,
                                n_estimators=best_ne)
best_rf.fit(X_train, y_train.ravel())

X_last10 = X_predict[-1:]
y_last10_pred = best_rf.predict(X_last10)
MSFT_pred = y_last10_pred[0]
print("For MSFT, the predicted percentage of price change on 2024.05.24 is", MSFT_pred)

MSFT_real = y_predict[-1]
print("For MSFT, the real percentage of price change on 2024.05.24 is", MSFT_real)

For MSFT, the predicted percentage of price change on 2024.05.24 is 0.056742276887312836
For MSFT, the real percentage of price change on 2024.05.24 is [0.7400477]


In [89]:
# AAPL

new_stock = "AAPL"
train, valid, test, predict = data_generator(new_stock)
X_train, y_train = tsv_generator(train)
X_predict, y_predict = tsv_generator(predict)

best_rf.fit(X_train, y_train)

X_last10 = X_predict[-1:]
y_last10_pred = best_rf.predict(X_last10)

AAPL_pred = y_last10_pred[0]
print("For " + new_stock + ", the predicted percentage of price change on 2024.05.24 is", AAPL_pred)

AAPL_real = y_predict[-1]
print("For " + new_stock + ", the real percentage of price change on 2024.05.24 is", AAPL_real)

<ipython-input-89-90e63383bb30>:8: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  best_rf.fit(X_train, y_train)


For AAPL, the predicted percentage of price change on 2024.05.24 is 0.1132373033698522
For AAPL, the real percentage of price change on 2024.05.24 is [1.65881355]


In [90]:
# V

new_stock = "V"
train, valid, test, predict = data_generator(new_stock)
X_train, y_train = tsv_generator(train)
X_predict, y_predict = tsv_generator(predict)

best_rf.fit(X_train, y_train)

X_last10 = X_predict[-1:]
y_last10_pred = best_rf.predict(X_last10)

V_pred = y_last10_pred[0]
print("For " + new_stock + ", the predicted percentage of price change on 2024.05.24 is", V_pred)

V_real = y_predict[-1]
print("For " + new_stock + ", the real percentage of price change on 2024.05.24 is", V_real)

<ipython-input-90-4189073cc654>:8: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  best_rf.fit(X_train, y_train)


For V, the predicted percentage of price change on 2024.05.24 is 0.13978967098247674
For V, the real percentage of price change on 2024.05.24 is [0.09480335]


**Training and hyperparameter tuning on Index data**

In [91]:
train, valid, test, predict = data_generator("^DJI")
X_train, y_train = tsv_generator(train)
X_valid, y_valid = tsv_generator(valid)
X_test, y_test = tsv_generator(test)
X_predict, y_predict = tsv_generator(predict)

In [92]:
rf_rmse = time_series_valid_test(X_train, y_train, 3, "valid")
ind = 0
min_rmse = 10
for max_depth in max_depth_list:
  for alpha in alpha_list:
        for n_estimators in n_estimators_list:
            if rf_rmse[ind] < min_rmse:
                min_rmse = rf_rmse[ind]
                best_md = max_depth
                best_alpha = alpha
                best_ne = n_estimators
            print(["(max_depth, alpha, n_estimators):",[max_depth, alpha, n_estimators]])
            print(rf_rmse[ind])
            ind += 1

['(max_depth, alpha, n_estimators):', [1, 0.1, 10]]
1.5937445304053337
['(max_depth, alpha, n_estimators):', [1, 0.1, 50]]
1.5937124778189207
['(max_depth, alpha, n_estimators):', [1, 0.1, 100]]
1.5936962603590292
['(max_depth, alpha, n_estimators):', [1, 0.1, 200]]
1.593701469986552
['(max_depth, alpha, n_estimators):', [1, 0.1, 400]]
1.5937104757100082
['(max_depth, alpha, n_estimators):', [1, 0.001, 10]]
1.606538425677974
['(max_depth, alpha, n_estimators):', [1, 0.001, 50]]
1.589065145705531
['(max_depth, alpha, n_estimators):', [1, 0.001, 100]]
1.5937033705370978
['(max_depth, alpha, n_estimators):', [1, 0.001, 200]]
1.5964458064119398
['(max_depth, alpha, n_estimators):', [1, 0.001, 400]]
1.5942826935216914
['(max_depth, alpha, n_estimators):', [1, 1e-05, 10]]
1.606538425677974
['(max_depth, alpha, n_estimators):', [1, 1e-05, 50]]
1.589065145705531
['(max_depth, alpha, n_estimators):', [1, 1e-05, 100]]
1.5937033705370978
['(max_depth, alpha, n_estimators):', [1, 1e-05, 200]]
1.59

In [93]:
# validation rmse
# hyperparameters:
print(min_rmse, best_md, best_alpha, best_ne)
min(rf_rmse)

1.589065145705531 1 0.001 50


1.589065145705531

In [94]:
# test rmse
rf_rmse, y_test_rf = time_series_valid_test(X_test, y_test, 3, "test", [best_md, best_alpha, best_ne])
rf_rmse

0.6744007160244486

In [95]:
best_rf_index = RandomForestRegressor(random_state=42, max_depth=best_md, ccp_alpha=best_alpha,
                                n_estimators=best_ne)
best_rf_index.fit(X_train, y_train.ravel())

RandomForestRegressor(ccp_alpha=0.001, max_depth=1, n_estimators=50,
                      random_state=42)

**Predicting the percantage change of the Index**

In [96]:
X_last10 = X_predict[-1:]
y_last10_pred = best_rf_index.predict(X_last10)
index_pred = y_last10_pred[0]
print("The predicted percentage of price change of the Dow Jones Index on 2024.05.24 is", index_pred)

index_real = y_predict[-1]
print("The predicted percentage of price change of the Dow Jones Index on 2024.05.24 is", index_real)

The predicted percentage of price change of the Dow Jones Index on 2024.05.24 is 0.03739496402891866
The predicted percentage of price change of the Dow Jones Index on 2024.05.24 is [0.01107922]


**Comparing each stock and the Index**

In [97]:
# MSFT

predict_diff = index_pred - MSFT_pred
real_diff = index_real - MSFT_real
error = np.abs(real_diff - predict_diff)
print("The predicted difference between the Index and the stock price change is", predict_diff)
print("The real difference between the Index and the stock price change is", real_diff)
print("The absolute value of the distance between real and predict is", error)

The predicted difference between the Index and the stock price change is -0.019347312858394174
The real difference between the Index and the stock price change is [-0.72896848]
The absolute value of the distance between real and predict is [0.70962117]


In [98]:
# AAPL

predict_diff = index_pred - AAPL_pred
real_diff = index_real - AAPL_real
error = np.abs(real_diff - predict_diff)
print("The predicted difference between the Index and the stock price change is", predict_diff)
print("The real difference between the Index and the stock price change is", real_diff)
print("The absolute value of the distance between real and predict is", error)

The predicted difference between the Index and the stock price change is -0.07584233934093354
The real difference between the Index and the stock price change is [-1.64773433]
The absolute value of the distance between real and predict is [1.57189199]


In [99]:
# V

predict_diff = index_pred - V_pred
real_diff = index_real - V_real
error = np.abs(real_diff-predict_diff)
print("The predicted difference between the Index and the stock price change is", predict_diff)
print("The real difference between the Index and the stock price change is", real_diff)
print("The absolute value of the distance between real and predict is", error)

The predicted difference between the Index and the stock price change is -0.10239470695355808
The real difference between the Index and the stock price change is [-0.08372414]
The absolute value of the distance between real and predict is [0.01867057]


# 6. Results Comparison *Charlie Yin*

| Model | Predicted Difference                             | Real Difference                                    | Absolute Distance                               |
|-------|--------------------------------------------------|----------------------------------------------------|-------------------------------------------------|
| RNN   | MSFT: 0.16339877 V: 0.15415367 AAPL: 0.14622732  | MSFT: -0.72896848 V: -0.08372414 AAPL: -1.64773433 | MSFT: 0.892367 V: 0.23787781 AAPL: 1.79396165   |
| SVM   | MSFT: -0.0550819 V: -0.1392939 AAPL:  -0.0157989 | MSFT: -0.72896848 V: -0.08372414 AAPL: -1.64773433 | MSFT: 0.67388656 V: 0.05556979 AAPL: 1.63193542 |
| RF    | MSFT: -0.0193473 V: -0.1023947 AAPL: -0.0758423  | MSFT: -0.72896848 V: -0.08372414 AAPL: -1.64773433 | MSFT: 0.70962117 V: 0.01867057 AAPL: 1.57189199 |

# 7. Conclusion *Charlie + Saya*

For the RNN-LSTM model:
We predicted the price change of MSFT (Microsoft) will be 0.0983, and the actual price change is 0.74. The direction is correct and the value difference is a little bit large.

We predicted the price change of V (Visa) is around 0.1075, and the actual price change is around 0.09848. The direction is correct and the value difference is smaller.

We predicted the price change of AAPL (Apple) is around 0.11548 and the actual price change of AAPL is around 1.6588. The direction is correct, but the value difference is kind large.

We predicted the Index price change is around 0.2617, and the actual price change is around 0.011, the direction is correct and the value difference is also a little bit large.

We predicted that the Index price change would be higher than the MSFT price change by 0.16339, but actually the stock outperform the index. The direction is not correct and the value difference is 0.63398.

We predicted that the index price change will be higher than the V price change, but actually V outperform the index. The direction is not correct and the difference is also larger.

We predicted that the index price change will be larger than the AAPL price change, but actually AAPL outperform the index, and difference is larger.

In conclusion, for RNN-LSTM model, on stock price change predictions, it does well on directions and not bad on values. But for the outperformance predictions, it does not do very well.

For SVM model

We predicted the price change of MSFT (Microsoft) will be 0.11302, and the actual price change is 0.74. The direction is correct and the value difference is a little bit large.

We predicted the price change of V (Visa) is around 0.197228, and the actual price change is around 0.09848. The direction is correct and the value difference is smaller.

We predicted the price change of AAPL (Apple) is around 0.073732 and the actual price change of AAPL is around 1.6588. The direction is correct, but the value difference is kind large.

We predicted the Index price change is around 0.057934, and the actual price change is around 0.011, the direction is correct and the value difference is reasonable.

We predicted that the MSFT Stock price change would be higher than the Index price change by 0.05508, but the actual difference is 0.72896, which is larger. The direction is correct and the value difference is 0.67388.

We predicted that the Stock price change will be higher than the Index change, and it is true.

We predicted that the AAPL stock price change will be larger than the Index price change, and actually AAPL outperforms the index.

In conclusion, for SVM model, on stock price change predictions, it does well on directions and not bad on values. For the outperformance predictions, its performance is in a reasonable range and the signs for prediction is correct.

For Random Forest Model:

We predicted that the price change of MSFT will be around 0.0567, it is actually 0.74, so the direction is correct and value difference is large.

We predicted that the price change of AAPL is 0.11, but it actually 1.6588, so the direction is correct, but the value difference is larger.

We predicted that the price change of V is 0.1397, and it is actually 0.0948, the direction is correct, and the value difference is smaller.

We predicted that the price change of the Index is 0.03739, and it’s actually 0.011, the direction is correct, and the value difference is good.

We predicted that MSFT will outperform by 0.019, and it’s actually that the MSFT does not outperform the index. The value difference is large.

We predicted that AAPL will outperform the index by around 0.0758, but it actually outperformed y 1.6477, so the direction is correct, but the value difference is large.

We predicted that V will outperform by 0.1024, and it actually outperformed by 0.083, the direction is correct and the value difference is small.

In conclusion, the Random Forest Model performs good on predicting the direction of stock price changes but the value difference is larger. In terms of ouperformance, it can also do good job on predicting the direction, but the value difference is large for AAPL and MSFT.
